### PR TITLE
[ubsan] Fix misaligned reads

### PR DESCRIPTION
**Pull request by cdesouza-chromium**
It is undefined behaviour to do a misligned read in c++, and it appears that the data being read by struct is produced misaligned from the rust side itself.

This change avoids the type of cast we had in place, which was getting trapped by ubsan.

Resolves https://github.com/brave/brave-browser/issues/41434
Resolves https://github.com/brave/brave-browser/issues/41432

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/nft_metadata_fetcher.cc [hunk 1]
```diff
@@ -24,24 +26,15 @@
 
 namespace {
 
-std::optional<uint32_t> DecodeUint32(const std::vector<uint8_t>& input,
+std::optional<uint32_t> DecodeUint32(base::span<const uint8_t> input,
                                      size_t& offset) {
-  if (offset >= input.size() || input.size() - offset < sizeof(uint32_t)) {
+  if (input.size() < offset + sizeof(uint32_t)) {
```
Comment (supermassive) [inline components/brave_wallet/browser/nft_metadata_fetcher.cc:31]: Strictly speaking this might overflow now.

Comment (cdesouza-chromium) [reply components/brave_wallet/browser/nft_metadata_fetcher.cc]: Reverted.

File: components/brave_wallet/browser/nft_metadata_fetcher.cc [hunk 2]
```diff
@@ -24,24 +26,15 @@
...
     return std::nullopt;
   }
 
-  // Read bytes in little endian order.
-  base::span<const uint8_t> s =
-      base::make_span(input.begin() + offset, sizeof(uint32_t));
-  uint32_t uint32_le = *reinterpret_cast<const uint32_t*>(s.data());
-
+  auto value = input.subspan(offset).first<4u>();
   offset += sizeof(uint32_t);
```
Comment (supermassive) [inline components/brave_wallet/browser/nft_metadata_fetcher.cc:36]: nit: Let's have either `4u` or `sizeof(uint32_t)` for both lines
And same below

**Review by supermassive — APPROVED** [review]
wallet core lgtm

File: components/brave_wallet/browser/nft_metadata_fetcher.cc (hunk 2 above)
Comment (cdesouza-chromium) [reply components/brave_wallet/browser/nft_metadata_fetcher.cc]: Done.

**Pull request merged**